[Bug]: Inconsistent avro format function needed between FILE_LOAD and STORAGE_WRITE_API methods

### What happened?

When implementing BigQuery STORAGE_WRITE_API performance test, it is found that withAvroFormatFunction needs to be set differently:

For FILE_LOAD:
```
BigQueryIO.<byte[]>write()
    .withAvroFormatFunction(
            writeRequest -> {
              byte[] data = writeRequest.getElement();
              GenericRecord record = new GenericData.Record(writeRequest.getSchema());
              record.put("data", ByteBuffer.wrap(data));
              return record;
            })
```
See https://github.com/apache/beam/blob/44a17cbe429699160e689bd88935c32c6c2de6b2/sdks/java/io/bigquery-io-perf-tests/src/test/java/org/apache/beam/sdk/bigqueryioperftests/BigQueryIOIT.java#L169


For STORAGE_WRITE_API, however, it needs to be
```
BigQueryIO.<byte[]>write()
    .withAvroFormatFunction(
            writeRequest -> {
              byte[] data = writeRequest.getElement();
              GenericRecord record = new GenericData.Record(writeRequest.getSchema());
              record.put("data", data);
              return record;
            })
```

Using `ByteBuffer.wrap(data)` will throw this Exception at run time:
```
Caused by: java.lang.ClassCastException: class java.nio.HeapByteBuffer cannot be cast to class [B (java.nio.HeapByteBuffer and [B are in module java.base of loader 'bootstrap')
        at org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.lambda$static$4(AvroGenericRecordToStorageApiProto.java:90)
        at org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.scalarToProtoValue(AvroGenericRecordToStorageApiProto.java:375)
        at org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.toProtoValue(AvroGenericRecordToStorageApiProto.java:332)
        at org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.toProtoValue(AvroGenericRecordToStorageApiProto.java:318)
        at org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.messageValueFromGenericRecordValue(AvroGenericRecordToStorageApiProto.java:296)
        at org.apache.beam.sdk.io.gcp.bigquery.AvroGenericRecordToStorageApiProto.messageFromGenericRecord(AvroGenericRecordToStorageApiProto.java:183)
        at org.apache.beam.sdk.io.gcp.bigquery.StorageApiDynamicDestinationsGenericRecord$GenericRecordConverter.toMessage(StorageApiDynamicDestinationsGenericRecord.java:73)
        at org.apache.beam.sdk.io.gcp.bigquery.StorageApiConvertMessages$ConvertMessagesDoFn.processElement(StorageApiConvertMessages.java:139)
```
at here:

https://github.com/apache/beam/blob/44a17cbe429699160e689bd88935c32c6c2de6b2/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/AvroGenericRecordToStorageApiProto.java#L90

In contrast, if directly feed a byte[] for FILE_LOAD, it gives the following error:

```
Caused by: org.apache.avro.file.DataFileWriter$AppendWriteException: org.apache.avro.AvroRuntimeException: Unknown datum type [B: [B@6f10f22c
        at org.apache.avro.file.DataFileWriter.append(DataFileWriter.java:317)
        at org.apache.beam.sdk.io.gcp.bigquery.AvroRowWriter.write(AvroRowWriter.java:52)
        at org.apache.beam.sdk.io.gcp.bigquery.WriteBundlesToFiles.processElement(WriteBundlesToFiles.java:236)
Caused by: org.apache.avro.AvroRuntimeException: Unknown datum type [B: [B@6f10f22c
        at org.apache.avro.generic.GenericData.getSchemaName(GenericData.java:933)
        at org.apache.avro.generic.GenericData.resolveUnion(GenericData.java:892)
        at org.apache.avro.generic.GenericDatumWriter.resolveUnion(GenericDatumWriter.java:307)
```

Ideally we should support ByteBuffer in AvroGenericRecordToStorageApiProto.

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BeamRowToStorageApiProto.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BeamRowToStorageApiProto.java
@@ -24,6 +24,7 @@ import com.google.protobuf.Descriptors.Descriptor;
 import com.google.protobuf.Descriptors.FieldDescriptor;
 import com.google.protobuf.DynamicMessage;
 import java.math.BigDecimal;
+import java.nio.ByteBuffer;
 import java.time.Instant;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
@@ -102,9 +103,23 @@ public class BeamRowToStorageApiProto {
           .put(TypeName.BOOLEAN, Function.identity())
           // A Beam DATETIME is actually a timestamp, not a DateTime.
           .put(TypeName.DATETIME, o -> ((ReadableInstant) o).getMillis() * 1000)
-          .put(TypeName.BYTES, o -> ByteString.copyFrom((byte[]) o))
+          .put(TypeName.BYTES, BeamRowToStorageApiProto::toProtoByteString)
           .put(TypeName.DECIMAL, o -> serializeBigDecimalToNumeric((BigDecimal) o))
           .build();
+
+  private static ByteString toProtoByteString(Object o) {
+    if (o instanceof byte[]) {
+      return ByteString.copyFrom((byte[]) o);
+    } else if (o instanceof ByteBuffer) {
+      return ByteString.copyFrom((ByteBuffer) o);
+    } else if (o instanceof String) {
+      return ByteString.copyFromUtf8((String) o);
+    } else {
+      throw new ClassCastException(
+          String.format(
+              "Cannot cast %s to a compatible object to build ByteString.", o.getClass()));
+    }
+  }
 
   // A map of supported logical types to their encoding functions.
   static final Map<String, BiFunction<LogicalType<?, ?>, Object, Object>> LOGICAL_TYPE_ENCODERS =

--- a/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BeamRowToStorageApiProtoTest.java
+++ b/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BeamRowToStorageApiProtoTest.java
@@ -28,6 +28,7 @@ import com.google.protobuf.Descriptors.Descriptor;
 import com.google.protobuf.Descriptors.FieldDescriptor;
 import com.google.protobuf.DynamicMessage;
 import java.math.BigDecimal;
+import java.nio.ByteBuffer;
 import java.nio.charset.StandardCharsets;
 import java.time.LocalDate;
 import java.time.LocalDateTime;
@@ -44,16 +45,14 @@ import org.apache.beam.sdk.values.Row;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Functions;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableList;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.commons.math3.util.Pair;
 import org.joda.time.Instant;
 import org.junit.Test;
 import org.junit.runner.RunWith;
 import org.junit.runners.JUnit4;
 
-@RunWith(JUnit4.class)
-@SuppressWarnings({
-  "nullness" // TODO(https://github.com/apache/beam/issues/20497)
-})
 /** Unit tests form {@link BeamRowToStorageApiProto}. */
+@RunWith(JUnit4.class)
 public class BeamRowToStorageApiProtoTest {
   private static final EnumerationType TEST_ENUM =
       EnumerationType.create("ONE", "TWO", "RED", "BLUE");
@@ -391,5 +390,29 @@ public class BeamRowToStorageApiProtoTest {
             .collect(Collectors.toMap(FieldDescriptor::getName, Functions.identity()));
     DynamicMessage nestedMsg = (DynamicMessage) msg.getField(fieldDescriptors.get("nested"));
     assertBaseRecord(nestedMsg);
+  }
+
+  @Test
+  public void testScalarToProtoValue() {
+    Map<FieldType, Iterable<Pair<Object, Object>>> testCases =
+        ImmutableMap.<FieldType, Iterable<Pair<Object, Object>>>builder()
+            .put(
+                FieldType.BYTES,
+                ImmutableList.of(
+                    Pair.create(BYTES, ByteString.copyFrom(BYTES)),
+                    Pair.create(ByteBuffer.wrap(BYTES), ByteString.copyFrom(BYTES)),
+                    Pair.create(
+                        new String(BYTES, StandardCharsets.UTF_8), ByteString.copyFrom(BYTES))))
+            .build();
+    for (Map.Entry<FieldType, Iterable<Pair<Object, Object>>> entry : testCases.entrySet()) {
+      entry
+          .getValue()
+          .forEach(
+              p -> {
+                assertEquals(
+                    p.getValue(),
+                    BeamRowToStorageApiProto.scalarToProtoValue(entry.getKey(), p.getKey()));
+              });
+    }
   }
 }

